Signal/src/util/AppUpdateNag.m

```diff
@@ -0,0 +1,130 @@
+//
+//  Copyright (c) 2017 Open Whisper Systems. All rights reserved.
+//
+
+#import "AppUpdateNag.h"
+#import "RegistrationViewController.h"
+#import "Signal-Swift.h"
+#import <ATAppUpdater/ATAppUpdater.h>
+#import <SignalServiceKit/TSStorageManager.h>
+
+NSString *const TSStorageManagerAppUpgradeNagCollection = @"TSStorageManagerAppUpgradeNagCollection";
+NSString *const TSStorageManagerAppUpgradeNagDate = @"TSStorageManagerAppUpgradeNagDate";
+
+@interface AppUpdateNag () <ATAppUpdaterDelegate>
+
+@property (nonatomic, readonly) TSStorageManager *storageManager;
+
+@end
+
+#pragma mark -
+
+@implementation AppUpdateNag
+
++ (instancetype)sharedInstance
+{
+    static AppUpdateNag *sharedInstance = nil;
+    static dispatch_once_t onceToken;
+    dispatch_once(&onceToken, ^{
+        sharedInstance = [[self alloc] initDefault];
+    });
+    return sharedInstance;
+}
+
+- (instancetype)initDefault
+{
+    TSStorageManager *storageManager = [TSStorageManager sharedManager];
+
+    return [self initWithStorageManager:storageManager];
+}
+
+- (instancetype)initWithStorageManager:(TSStorageManager *)storageManager
+{
+    self = [super init];
+
+    if (!self) {
+        return self;
+    }
+
+    OWSAssert(storageManager);
+
+    _storageManager = storageManager;
+
+    OWSSingletonAssert();
+
+    return self;
+}
+
+- (void)showAppUpgradeNagIfNecessary
+{
+    // Only show nag if we are "at rest" in the home view or registration view without any
+    // alerts or dialogs showing.
+    UIViewController *frontmostViewController =
+    [UIApplication sharedApplication].frontmostViewController;
+    OWSAssert(frontmostViewController);
+    BOOL canPresent = ([frontmostViewController isKindOfClass:[SignalsViewController class]] ||
+                       [frontmostViewController isKindOfClass:[RegistrationViewController class]]);
+    if (!canPresent) {
+        return;
+    }
+    
+    NSDate *lastNagDate = [[TSStorageManager sharedManager] dateForKey:TSStorageManagerAppUpgradeNagDate
+                                                          inCollection:TSStorageManagerAppUpgradeNagCollection];
+    const NSTimeInterval kMinute = 60.f;
+    const NSTimeInterval kHour = 60 * kMinute;
+    const NSTimeInterval kDay = 24 * kHour;
+    const NSTimeInterval kNagFrequency = kDay * 14;
+    BOOL canNag = (!lastNagDate || fabs(lastNagDate.timeIntervalSinceNow) > kNagFrequency);
+    if (!canNag) {
+        return;
+    }
+
+    // NOTE: The iTunes app store API exposes "short" version numbers, so
```

Comment:
Surprised to hear this - Do you have any more details? Even if they don't know our build number, we are required to increment the version with every release.

Doing a little snooping - the output of`curl https://itunes.apple.com/lookup?id=874139669`

includes the string:

```
"version":"2.13.1
```

So it seems like sufficient information exists to determine our hotfix release - but I haven't dug into ATAppUpdater to verify they are using that.

However, it seems like this might not work to notify our testflight users that there is a new build available for their currently Version (but testflight notifications mostly take care of that anyway).